Lifecycle Methods and Event Alternative to `sessionsId` (#285)
ChainAgnostic/CAIPs: CAIPs/caip-285.md

---
caip: 285
title: JSON-RPC Method for Revoking Session Authorizations
author: [Alex Donesky] (@adonesky1)
discussions-to: TBD
status: Draft
type: Standard
created: 2024-07-12
requires: 25, 217
---

## Simple Summary

CAIP-285 introduces the `wallet_revokeSession` method for fully revoking the authorizations and properties of an active [CAIP-25][] session.

## Abstract

This proposal aims to enhance session management for [CAIP-25][] initiated sessions by defining a new JSON-RPC method for revoking sessions. This method provides an explicit protocol for revoking sessions with or without `sessionId`s.

## Motivation

The motivation behind this proposal is to enhance the flexibility of [CAIP-25][] initated sessions by enabling the revocation of session authorizations without `sessionId`s, which don't map well to extension-based wallet's dapp connections and could add unnecessary constraints and burdens to existing flows. The proposed method provides an intuitive way to revoke authorizations of an active session, simplifying the management of session lifecycles.

## Specification

### Definition

> The `wallet_revokeSession` method revokes the entire active session. If a `sessionId` is provided, it revokes that specific session; otherwise, it revokes the single active session between the wallet and the caller

Reviewer: ```suggestion
The `wallet_revokeSession` method revokes the entire active session.
If a `sessionId` is provided, it revokes that specific session only; if none is provide and there is only one session to revoke, it gets revokes and a success result is returned; otherwise, an appropriate error message is sent.
```

Author: Seems right except I would say:
```
if none is provided and there is an active session without a `sessionId`, it gets revokes and a success result is returned;
```
Since `if none is provided and there is only one session to revoke` seems to me to make ambiguous whether there can be multiple active sessions without `sessionIds`

Author: updated with some hybrid of our language here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/405c5a03d7d0386c4f3d87bcf8db8de81af6257f

**Parameters:**

- `sessionId` (string, optional): The session identifier.

### Request

> The caller would interface with a wallet via the same provider by which it called `wallet_createSession` to revoke a session by calling the following JSON-RPC request:

Reviewer: is it always a `provider` or would something more generic like `channel` be more inclusive of new possibilities?

Author: channel makes sense!

Author: done here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/405c5a03d7d0386c4f3d87bcf8db8de81af6257f

```jsonc
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_revokeSession",
  "params": {}
}
```

### Response

The wallet can respond to this method with either a success result or an error message.

### Success

Upon a successful `wallet_revokeSession` call a wallet should remove authorizations and session properties associated with the revoked session.

An example of a successful response follows:

```jsonc
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": true
}
```

### Failure States

The response MUST NOT be a JSON-RPC success result in any of the following failure states.

#### Generic Failure Code

Unless the dapp is known to the wallet and trusted, the generic/undefined error response:

```jsonc
{
  "id": 1,
  "jsonrpc": "2.0",
  "error": {
    "code": 0,
    "message": "Unknown error"
  }
}
```

is RECOMMENDED for any of the following cases:

- a sessionId is passed but not recognized,

> - no sessionId is passed and only active session(s) have sessionIds,

Reviewer: ```suggestion
- no sessionId is passed and only active session(s) have sessionIds, or
```

Author: https://github.com/ChainAgnostic/CAIPs/commit/405c5a03d7d0386c4f3d87bcf8db8de81af6257f

> - there are no active sessions,

Reviewer: ```suggestion
- there are no active sessions
```

Author: https://github.com/ChainAgnostic/CAIPs/commit/405c5a03d7d0386c4f3d87bcf8db8de81af6257f

---
caip: 285
title: JSON-RPC Method for Revoking Session Authorizations
author: [Alex Donesky] (@adonesky1)

> discussions-to: TBD

Reviewer: ```suggestion
discussions-to: https://github.com/ChainAgnostic/CAIPs/pull/285/files
```

If a `sessionId` parameter is provided, it revokes that specific session only;
if no `sessionId` parameter is provided and there is an active session without a `sessionId` this session gets revoked and a success result is returned;
otherwise, an appropriate error message is sent.

**Parameters:**

- `sessionId` (string, optional): The session identifier.

### Request

The caller would interface with a wallet via the same channel by which it called `wallet_createSession` to revoke a session by calling the following JSON-RPC request:

```jsonc
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_revokeSession",
  "params": {}
}
```

### Response

The wallet can respond to this method with either a success result or an error message.

### Success

Upon a successful `wallet_revokeSession` call a wallet should remove authorizations and session properties associated with the revoked session.

An example of a successful response follows:

```jsonc
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": true
}
```

### Failure States

The response MUST NOT be a JSON-RPC success result in any of the following failure states.

#### Generic Failure Code

> Unless the dapp is known to the wallet and trusted, the generic/undefined error response:

Reviewer 2: Should be further defined in subsequent PRs what the threshold for trusted is because otherwise this leaves this part of the spec quite ambiguous for a wallet to implement

Reviewer: agreed, this has been a TODO that's been nagging on me since Kyle first recommended doing (as Aaron recommended on the more recent call) as little error-handling as possible 😅 .  i opened a tracking issue just now so it can nag everyone, not just me.
